tsl/src/continuous_aggs/create.c

```diff
@@ -707,6 +838,54 @@ cagg_create(const CreateTableAsStmt *create_stmt, ViewStmt *stmt, Query *panquer
 	 */
 	stmt->options = NULL;
 
+	/*
+	 * If we want to create a continuous aggregate using WAL-based
+	 * invalidation collection, it should not have an invalidation trigger
+	 * already.
+	 */
+	if (collect_using == ContinuousAggCollectUsingWal &&
+		has_invalidation_trigger(bucket_info->htoid))
+	{
+		ereport(ERROR,
+				errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+				errmsg("cannot use WAL-based invalidation collection with hypertable that is "
+					   "using trigger-based invalidation collection"));
+	}
+
+	/*
+	 * If we want to create a continuous aggregate with a trigger, the
+	 * hypertable should either have a trigger already, or if there is no
+	 * trigger, no hypertables should be attached to it.
+	 */
+	if (collect_using == ContinuousAggCollectUsingTrigger &&
+		!has_invalidation_trigger(bucket_info->htoid) &&
+		number_of_continuous_aggs_attached(bucket_info->htid) > 0)
+		ereport(ERROR,
+				errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+				errmsg("cannot use trigger-based invalidation collection with hypertable that is "
+					   "using wal-based invalidation collection"));
+
+	/*
+	 * If we didn't specify which collection method to use, we pick whatever
+	 * is used, or the default.
+	 */
+	if (collect_using == ContinuousAggCollectUsingDefault)
+	{
+		if (number_of_continuous_aggs_attached(bucket_info->htid) == 0)
+			collect_using = ContinuousAggCollectUsingTrigger; /* default case */
+		else if (has_invalidation_trigger(bucket_info->htoid))
+			collect_using = ContinuousAggCollectUsingTrigger;
+		else
+			collect_using = ContinuousAggCollectUsingWal;
+	}
+
+	/*
+	 * Step 0: Create logical replication slot if used. This needs to be first
+	 * since we cannot do this in a transaction that has performed any writes.
+	 */
+	if (collect_using == ContinuousAggCollectUsingWal)
+		cagg_add_logical_decoding_slot();
+
```

Comment:
Need to add a test to check what happens if we roll back a transaction creating a continuous aggregate.